synapse/storage/databases/main/deviceinbox.py

```diff
@@ -252,28 +261,32 @@ async def _get_device_messages(
                     there may be more messages to retrieve. If `limit` is not set, then this
                     is always equal to 'to_stream_id'.
         """
-        # A limit can only be applied when querying for a single user ID / device ID tuple.
-        if limit:
-            if not device_ids:
+        if not user_ids:
+            logger.warning("No users provided upon querying for device IDs")
+            return {}, to_stream_id
+
+        if len(user_ids) > 1:
+            if device_id is not None:
                 raise AssertionError(
-                    "Programming error: _get_new_device_messages was passed 'limit' "
-                    "but not device_ids. This could lead to querying multiple user ID "
-                    "/ device ID pairs, which is not compatible with 'limit'"
+                    "Programming error: 'device_id' cannot be supplied to "
+                    "_get_device_messages when >1 user_id has been provided"
                 )
 
-            if len(user_ids) > 1 or len(device_ids) > 1:
+            # A limit can only be applied when querying for a single user ID / device ID tuple.
+            if limit:
```

Comment:
nb that a limit of zero is different from a limit of `None`

```suggestion
            if limit is not None:
```